TorchSharp.PyBridge/PyBridgeModuleExtensions.cs

```diff
@@ -116,35 +114,107 @@ public static class PyBridgeModuleExtensions {
+        /// dictionary has been fully loaded.
         /// </remarks>
         public static Module load_py(this Module module, System.IO.Stream stream, bool strict = true, IList<string>? skip = null, Dictionary<string, bool>? loadedParameters = null, bool leaveOpen = false) {
-            // Create a dispose score so that we don't keep anyof the loaded tensors past this function
+            // Create a dispose score so that we don't keep any of the loaded tensors past this function
             using var d = torch.NewDisposeScope();
             using var d2 = torch.no_grad(); // To circumvent a bug introduced in 0.102.0
 
-            // Unpickle the state dictionary into memory
-            var stateHashtable = PyTorchUnpickler.UnpickleStateDict(stream, leaveOpen);
+            // Unpickle the state dictionary into memory.
+            // Keep stream open because tensors will not get deserialized yet.
+            var unpickled = PyTorchUnpickler.UnpickleStateDict(stream, leaveOpen: true, skipTensorRead: true);
 
             // Convert the hashtable to a dictionary of string->tensor
-            var stateDict = new Dictionary<string, torch.Tensor>();
-            foreach (string key in stateHashtable.Keys)
-                stateDict.Add(key, (torch.Tensor)stateHashtable[key]!);
+            Dictionary<string, PyTorchUnpickler.TensorConstructorArgs> unpickledConstructors = new();
 
-            // Load it in using the builtin function
-            var (_, unexpectedKeys) = module.load_state_dict(stateDict, strict, skip);
+            foreach (string key in unpickled.Keys) {
+                unpickledConstructors.Add(key, (PyTorchUnpickler.TensorConstructorArgs)unpickled[key]!);
+            }
 
-            // Fill in the loadedParameters dictionary, if relevant
-            if (loadedParameters is not null) {
-                foreach (string key in stateDict.Keys)
-                    loadedParameters[key] = true;
-                foreach (string key in unexpectedKeys)
-                    loadedParameters[key] = false;
+            var (_, unexpectedKeys) = load_state_dict(module, unpickledConstructors, strict, skip);
+
+            // Close stream now that tensor streams have been read.
+            stream.Close ();
+
+            if (loadedParameters is null) {
+                return module;
+            }
+
+            // Fill in the loadedParameters dictionary
+            foreach (var key in unpickledConstructors.Keys) {
+                loadedParameters[key] = true;
+            }
+
+            foreach (var key in unexpectedKeys) {
+                loadedParameters[key] = false;
             }
 
             return module;
         }
 
+        /// <summary>
+        /// Mirrors the implementation of module.load_state_dict but performs tensor reading
+        /// with less intermediate memory overhead.
+        /// </summary>
+        static (IList<string> missing_keys, IList<string> unexpected_keyes) load_state_dict(
```

Comment:
Once the implementation is copied and only used internally, we can fix the `keyes` spelling mistake.

Reply:
Fixed.  Worth mentioning that `missing_keys` is not being used in the current implementation.  I kept it only to preserve the legacy signature, but it could be cleaned up.  If we do it, tough, then the semantics of the single returned list become ambiguous.

Reply:
Right, good point.